File: native/Sources/Isometry/Bridge/Optimization/BinarySerializer.swift
/**
 * BinarySerializer for WebView Bridge Optimization
 *
 * Implements MessagePack binary serialization for Swift with Codable integration,
 * performance tracking, and error handling. Provides 40-60% payload reduction
 * compared to JSON baseline with type-safe encoding/decoding.
 */

import Foundation
import QuartzCore
import MessagePack
import OSLog

/// Configuration options for binary serialization
public struct SerializationOptions: Sendable {
    public let enableCompression: Bool
    public let maxDepth: Int
    public let enableTypeValidation: Bool

    public init(
        enableCompression: Bool = false,
        maxDepth: Int = 10,
        enableTypeValidation: Bool = true
    ) {
        self.enableCompression = enableCompression
        self.maxDepth = maxDepth
        self.enableTypeValidation = enableTypeValidation
    }
}

/// Result of serialization operation with performance metrics
public struct SerializationResult: Sendable {
    public let data: Data
    public let originalSize: Int
    public let compressedSize: Int
    public let compressionRatio: Double
    public let serializationTime: TimeInterval

    public init(
        data: Data,
        originalSize: Int,
        compressedSize: Int,
        compressionRatio: Double,
        serializationTime: TimeInterval
    ) {
        self.data = data
        self.originalSize = originalSize
        self.compressedSize = compressedSize
        self.compressionRatio = compressionRatio
        self.serializationTime = serializationTime
    }
}

/// Result of deserialization operation with performance metrics
public struct DeserializationResult<T>: Sendable where T: Sendable {
    public let data: T
    public let deserializationTime: TimeInterval
    public let dataSize: Int

    public init(data: T, deserializationTime: TimeInterval, dataSize: Int) {
        self.data = data
        self.deserializationTime = deserializationTime
        self.dataSize = dataSize
    }
}

/// Comprehensive metrics for serialization performance tracking
public struct SerializationMetrics: Sendable {
    public let totalSerialized: Int
    public let totalDeserialized: Int
    public let totalBytesSaved: Int
    public let averageCompressionRatio: Double
    public let averageSerializationTime: TimeInterval
    public let averageDeserializationTime: TimeInterval
    public let errorCount: Int

    public init(
        totalSerialized: Int,
        totalDeserialized: Int,
        totalBytesSaved: Int,
        averageCompressionRatio: Double,
        averageSerializationTime: TimeInterval,
        averageDeserializationTime: TimeInterval,
        errorCount: Int
    ) {
        self.totalSerialized = totalSerialized
        self.totalDeserialized = totalDeserialized
        self.totalBytesSaved = totalBytesSaved
        self.averageCompressionRatio = averageCompressionRatio
        self.averageSerializationTime = averageSerializationTime
        self.averageDeserializationTime = averageDeserializationTime
        self.errorCount = errorCount
    }
}

/// BinarySerializer provides MessagePack encoding/decoding with performance tracking
public final class BinarySerializer: Sendable {
    // MARK: - Properties

    private let options: SerializationOptions
    private let logger = Logger(subsystem: "com.isometry.bridge", category: "BinarySerializer")

    // Thread-safe metrics tracking
    private let metricsQueue = DispatchQueue(label: "com.isometry.binarySerializer.metrics")
    private var _metrics = SerializationMetrics(
        totalSerialized: 0,
        totalDeserialized: 0,
        totalBytesSaved: 0,
        averageCompressionRatio: 1.0,
        averageSerializationTime: 0,
        averageDeserializationTime: 0,
        errorCount: 0
    )

    // Performance tracking arrays (maintain last 100 measurements)
    private var serializationTimes: [TimeInterval] = []
    private var deserializationTimes: [TimeInterval] = []
    private var compressionRatios: [Double] = []

    // MARK: - Initialization

    /// Initialize BinarySerializer with configuration options
    /// - Parameter options: Serialization configuration options
    public init(options: SerializationOptions = SerializationOptions()) {
        self.options = options

        logger.debug("BinarySerializer initialized with maxDepth: \(options.maxDepth), validation: \(options.enableTypeValidation)")
    }

    // MARK: - Public Interface

    /// Serialize Codable data to MessagePack binary format
    /// - Parameter data: The data to serialize
    /// - Returns: Serialization result with performance metrics
    /// - Throws: BinarySerializationError if serialization fails
    public func serialize<T: Codable & Sendable>(_ data: T) throws -> SerializationResult {
        let startTime = CACurrentMediaTime()

        // Validate input if enabled
        if options.enableTypeValidation {
            try validateSerializationInput(data)
        }

        // Create signpost for performance tracking
        let signpostID = OSSignpostID(log: logger.osLog)
        os_signpost(.begin, log: logger.osLog, name: "serialize",
                   signpostID: signpostID, "Serializing data")

        do {
            // Calculate original JSON size for comparison
            let jsonData = try JSONEncoder().encode(data)
            let originalSize = jsonData.count

            // Serialize with MessagePack
            let binaryData = try MessagePackEncoder().encode(data)
            let compressedSize = binaryData.count

            let serializationTime = CACurrentMediaTime() - startTime
            let compressionRatio = Double(originalSize) / Double(compressedSize)

            // End signpost
            os_signpost(.end, log: logger.osLog, name: "serialize",
                       signpostID: signpostID, "Completed in %.2fms, ratio: %.2fx",
                       serializationTime * 1000, compressionRatio)

            // Update metrics on background queue
            updateSerializationMetrics(
                originalSize: originalSize,
                compressedSize: compressedSize,
                compressionRatio: compressionRatio,
                serializationTime: serializationTime
            )

            return SerializationResult(
                data: binaryData,
                originalSize: originalSize,
                compressedSize: compressedSize,
                compressionRatio: compressionRatio,
                serializationTime: serializationTime
            )

        } catch let error as EncodingError {
            incrementErrorCount()
            throw BinarySerializationError.serializationFailed("MessagePack encoding failed: \(error.localizedDescription)")
        } catch {
            incrementErrorCount()
            throw BinarySerializationError.serializationFailed("Serialization failed: \(error.localizedDescription)")
        }
    }

    /// Deserialize MessagePack binary data to Codable type
    /// - Parameters:
    ///   - binaryData: The MessagePack binary data
    ///   - type: The target Codable type
    /// - Returns: Deserialization result with performance metrics
    /// - Throws: BinarySerializationError if deserialization fails
    public func deserialize<T: Codable & Sendable>(_ binaryData: Data, as type: T.Type) throws -> DeserializationResult<T> {
        let startTime = CACurrentMediaTime()

        // Create signpost for performance tracking
        let signpostID = OSSignpostID(log: logger.osLog)
        os_signpost(.begin, log: logger.osLog, name: "deserialize",
                   signpostID: signpostID, "Deserializing data")

        do {
            // Deserialize with MessagePack
            let data = try MessagePackDecoder().decode(type, from: binaryData)

            let deserializationTime = CACurrentMediaTime() - startTime
            let dataSize = binaryData.count

            // Validate output if enabled
            if options.enableTypeValidation {
                validateDeserializationOutput(data)
            }

            // End signpost
            os_signpost(.end, log: logger.osLog, name: "deserialize",
                       signpostID: signpostID, "Completed in %.2fms",
                       deserializationTime * 1000)

            // Update metrics
            updateDeserializationMetrics(deserializationTime: deserializationTime)

            return DeserializationResult(
                data: data,
                deserializationTime: deserializationTime,
                dataSize: dataSize
            )

        } catch let error as DecodingError {
            incrementErrorCount()
            throw BinarySerializationError.deserializationFailed("MessagePack decoding failed: \(error.localizedDescription)")
        } catch {
            incrementErrorCount()
            throw BinarySerializationError.deserializationFailed("Deserialization failed: \(error.localizedDescription)")
        }
    }

    /// Get current serialization metrics (thread-safe)
    /// - Returns: Current metrics snapshot
    public func getMetrics() -> SerializationMetrics {
        return metricsQueue.sync { _metrics }
    }

    /// Reset metrics (useful for testing)
    public func resetMetrics() {
        metricsQueue.sync {
            _metrics = SerializationMetrics(
                totalSerialized: 0,
                totalDeserialized: 0,
                totalBytesSaved: 0,
                averageCompressionRatio: 1.0,
                averageSerializationTime: 0,
                averageDeserializationTime: 0,
                errorCount: 0
            )
            serializationTimes = []
            deserializationTimes = []
            compressionRatios = []
        }
    }

    // MARK: - Private Methods

    /// Validate serialization input
    private func validateSerializationInput<T>(_ data: T) throws {
        // Basic validation - can be extended
        if case Optional<Any>.none = data {
            throw BinarySerializationError.validationFailed("Cannot serialize nil data")
        }
    }

    /// Validate deserialization output
    private func validateDeserializationOutput<T>(_ data: T) {
        if case Optional<Any>.none = data {
            logger.warning("Deserialized data is nil")
        }
    }

    /// Update serialization metrics (thread-safe)
    private func updateSerializationMetrics(
        originalSize: Int,
        compressedSize: Int,
        compressionRatio: Double,
        serializationTime: TimeInterval
    ) {
        metricsQueue.async { [weak self] in
            guard let self = self else { return }

            let totalSerialized = self._metrics.totalSerialized + 1
            let totalBytesSaved = self._metrics.totalBytesSaved + (originalSize - compressedSize)

            // Track performance arrays (keep last 100 measurements)
            self.serializationTimes.append(serializationTime)
            self.compressionRatios.append(compressionRatio)

            if self.serializationTimes.count > 100 {
                self.serializationTimes.removeFirst()
            }
            if self.compressionRatios.count > 100 {
                self.compressionRatios.removeFirst()
            }

            // Calculate averages
            let averageSerializationTime = self.serializationTimes.reduce(0, +) / Double(self.serializationTimes.count)
            let averageCompressionRatio = self.compressionRatios.reduce(0, +) / Double(self.compressionRatios.count)

            self._metrics = SerializationMetrics(
                totalSerialized: totalSerialized,
                totalDeserialized: self._metrics.totalDeserialized,
                totalBytesSaved: totalBytesSaved,
                averageCompressionRatio: averageCompressionRatio,
                averageSerializationTime: averageSerializationTime,
                averageDeserializationTime: self._metrics.averageDeserializationTime,
                errorCount: self._metrics.errorCount
            )
        }
    }

    /// Update deserialization metrics (thread-safe)
    private func updateDeserializationMetrics(deserializationTime: TimeInterval) {
        metricsQueue.async { [weak self] in
            guard let self = self else { return }

            let totalDeserialized = self._metrics.totalDeserialized + 1

            self.deserializationTimes.append(deserializationTime)

            if self.deserializationTimes.count > 100 {
                self.deserializationTimes.removeFirst()
            }

            let averageDeserializationTime = self.deserializationTimes.reduce(0, +) / Double(self.deserializationTimes.count)

            self._metrics = SerializationMetrics(
                totalSerialized: self._metrics.totalSerialized,
                totalDeserialized: totalDeserialized,
                totalBytesSaved: self._metrics.totalBytesSaved,
                averageCompressionRatio: self._metrics.averageCompressionRatio,
                averageSerializationTime: self._metrics.averageSerializationTime,
                averageDeserializationTime: averageDeserializationTime,
                errorCount: self._metrics.errorCount
            )
        }
    }

    /// Increment error count (thread-safe)
    private func incrementErrorCount() {
        metricsQueue.async { [weak self] in
            guard let self = self else { return }

            self._metrics = SerializationMetrics(
                totalSerialized: self._metrics.totalSerialized,
                totalDeserialized: self._metrics.totalDeserialized,
                totalBytesSaved: self._metrics.totalBytesSaved,
                averageCompressionRatio: self._metrics.averageCompressionRatio,
                averageSerializationTime: self._metrics.averageSerializationTime,
                averageDeserializationTime: self._metrics.averageDeserializationTime,
                errorCount: self._metrics.errorCount + 1
            )
        }
    }
}

// MARK: - Error Types

/// Errors that can occur during binary serialization
public enum BinarySerializationError: LocalizedError, Sendable {
    case serializationFailed(String)
    case deserializationFailed(String)
    case validationFailed(String)

    public var errorDescription: String? {
        switch self {
        case .serializationFailed(let message):
            return "Serialization failed: \(message)"
        case .deserializationFailed(let message):
            return "Deserialization failed: \(message)"
        case .validationFailed(let message):
            return "Validation failed: \(message)"
        }
    }
}

// MARK: - Logger Extension

extension Logger {
    var osLog: OSLog {
        return OSLog(subsystem: subsystem, category: category)
    }
}

File: src/utils/bridge-optimization/binary-serializer.ts
/**
 * Binary Serializer for WebView Bridge Optimization
 *
 * Implements MessagePack binary serialization for 40-60% payload reduction
 * compared to JSON baseline with error handling and type safety.
 */

import { encode as pack, decode as unpack } from '@msgpack/msgpack';

export interface SerializationOptions {
  enableCompression?: boolean;
  maxDepth?: number;
  enableTypeValidation?: boolean;
}

export interface SerializationResult {
  data: Uint8Array;
  originalSize: number;
  compressedSize: number;
  compressionRatio: number;
  serializationTime: number;
}

export interface DeserializationResult<T = unknown> {
  data: T;
  deserializationTime: number;
  dataSize: number;
}

export interface SerializationMetrics {
  totalSerialized: number;
  totalDeserialized: number;
  totalBytesSaved: number;
  averageCompressionRatio: number;
  averageSerializationTime: number;
  averageDeserializationTime: number;
  errorCount: number;
}

/**
 * BinarySerializer provides MessagePack encoding/decoding for bridge messages
 * with performance tracking and error handling
 */
export class BinarySerializer {
  private readonly enableTypeValidation: boolean;
  private readonly maxDepth: number;

  // Metrics tracking
  private metrics: SerializationMetrics = {
    totalSerialized: 0,
    totalDeserialized: 0,
    totalBytesSaved: 0,
    averageCompressionRatio: 1.0,
    averageSerializationTime: 0,
    averageDeserializationTime: 0,
    errorCount: 0
  };

  // Performance tracking arrays
  private serializationTimes: number[] = [];
  private deserializationTimes: number[] = [];
  private compressionRatios: number[] = [];

  constructor(options: SerializationOptions = {}) {
    this.enableTypeValidation = options.enableTypeValidation ?? true;
    this.maxDepth = options.maxDepth ?? 10;
  }

  /**
   * Serialize data to MessagePack binary format
   */
  serialize<T = unknown>(data: T): SerializationResult {
    const startTime = performance.now();

    try {
      // Validate input if enabled
      if (this.enableTypeValidation) {
        this.validateSerializationInput(data);
      }

      // Calculate original JSON size for comparison
      const jsonString = JSON.stringify(data);
      const originalSize = new TextEncoder().encode(jsonString).length;

      // Serialize with MessagePack (using default options)
      const binaryData = pack(data);
      const compressedSize = binaryData.length;

      const serializationTime = performance.now() - startTime;
      const compressionRatio = originalSize / compressedSize;

      // Update metrics
      this.updateSerializationMetrics(originalSize, compressedSize, compressionRatio, serializationTime);

      return {
        data: binaryData,
        originalSize,
        compressedSize,
        compressionRatio,
        serializationTime
      };

    } catch (error) {
      this.metrics.errorCount++;
      throw new Error(`Serialization failed: ${error instanceof Error ? error.message : String(error)}`);
    }
  }

  /**
   * Deserialize MessagePack binary data back to original format
   */
  deserialize<T = unknown>(binaryData: Uint8Array): DeserializationResult<T> {
    const startTime = performance.now();

    try {
      // Deserialize with MessagePack (using default options)
      const data = unpack(binaryData) as T;

      const deserializationTime = performance.now() - startTime;
      const dataSize = binaryData.length;

      // Validate output if enabled
      if (this.enableTypeValidation) {
        this.validateDeserializationOutput(data);
      }

      // Update metrics
      this.updateDeserializationMetrics(deserializationTime);

      return {
        data,
        deserializationTime,
        dataSize
      };

    } catch (error) {
      this.metrics.errorCount++;
      throw new Error(`Deserialization failed: ${error instanceof Error ? error.message : String(error)}`);
    }
  }

  /**
   * Convenience method for encoding bridge messages
   */
  encodeBridgeMessage(message: {
    id: string;
    handler: string;
    method: string;
    params: Record<string, unknown>;
    timestamp: number;
  }): SerializationResult {
    return this.serialize(message);
  }

  /**
   * Convenience method for decoding bridge messages
   */
  decodeBridgeMessage(binaryData: Uint8Array): DeserializationResult<{
    id: string;
    handler: string;
    method: string;
    params: Record<string, unknown>;
    timestamp: number;
  }> {
    return this.deserialize(binaryData);
  }

  /**
   * Encode multiple messages as a batch
   */
  encodeBatch(messages: unknown[]): SerializationResult {
    const batchData = {
      type: 'batch',
      messages,
      batchSize: messages.length,
      timestamp: performance.now()
    };

    return this.serialize(batchData);
  }

  /**
   * Decode a batch of messages
   */
  decodeBatch<T = unknown>(binaryData: Uint8Array): DeserializationResult<{
    type: 'batch';
    messages: T[];
    batchSize: number;
    timestamp: number;
  }> {
    return this.deserialize(binaryData);
  }

  /**
   * Compare serialization efficiency with JSON
   */
  compareWithJSON<T = unknown>(data: T): {
    messagepack: SerializationResult;
    json: { size: number; time: number };
    efficiency: {
      sizeReduction: number;
      compressionRatio: number;
      isMoreEfficient: boolean;
    };
  } {
    // Serialize with MessagePack
    const messagepackResult = this.serialize(data);

    // Serialize with JSON for comparison
    const jsonStartTime = performance.now();
    const jsonString = JSON.stringify(data);
    const jsonSize = new TextEncoder().encode(jsonString).length;
    const jsonTime = performance.now() - jsonStartTime;

    const sizeReduction = ((jsonSize - messagepackResult.compressedSize) / jsonSize) * 100;
    const compressionRatio = jsonSize / messagepackResult.compressedSize;

    return {
      messagepack: messagepackResult,
      json: { size: jsonSize, time: jsonTime },
      efficiency: {
        sizeReduction,
        compressionRatio,
        isMoreEfficient: messagepackResult.compressedSize < jsonSize
      }
    };
  }

  /**
   * Validate serialization input
   */
  private validateSerializationInput<T>(data: T): void {
    if (data === null || data === undefined) {
      throw new Error('Cannot serialize null or undefined data');
    }

    // Check for circular references
    try {
      JSON.stringify(data);
    } catch (error) {
      if (error instanceof Error && error.message.includes('circular')) {
        throw new Error('Cannot serialize data with circular references');
      }
      throw error;
    }

    // Check depth (basic check)
    if (typeof data === 'object' && this.getObjectDepth(data) > this.maxDepth) {
      throw new Error(`Object depth exceeds maximum allowed depth of ${this.maxDepth}`);
    }
  }

  /**
   * Validate deserialization output
   */
  private validateDeserializationOutput<T>(data: T): void {
    if (data === null || data === undefined) {
      console.warn('[BinarySerializer] Deserialized data is null or undefined');
    }
  }

  /**
   * Calculate object depth (rough approximation)
   */
  private getObjectDepth(obj: unknown, depth = 0): number {
    if (depth > this.maxDepth) return depth;
    if (obj === null || typeof obj !== 'object') return depth;

    let maxChildDepth = depth;

    try {
      for (const value of Object.values(obj as Record<string, unknown>)) {
        const childDepth = this.getObjectDepth(value, depth + 1);
        maxChildDepth = Math.max(maxChildDepth, childDepth);
      }
    } catch {
      // If iteration fails, return current depth
      return depth;
    }

    return maxChildDepth;
  }

  /**
   * Update serialization metrics
   */
  private updateSerializationMetrics(
    originalSize: number,
    compressedSize: number,
    compressionRatio: number,
    serializationTime: number
  ): void {
    this.metrics.totalSerialized++;
    this.metrics.totalBytesSaved += (originalSize - compressedSize);

    // Track performance arrays (keep last 100 measurements)
    this.serializationTimes.push(serializationTime);
    this.compressionRatios.push(compressionRatio);

    if (this.serializationTimes.length > 100) {
      this.serializationTimes.shift();
    }
    if (this.compressionRatios.length > 100) {
      this.compressionRatios.shift();
    }

    // Update averages
    this.metrics.averageSerializationTime =
      this.serializationTimes.reduce((a, b) => a + b, 0) / this.serializationTimes.length;
    this.metrics.averageCompressionRatio =
      this.compressionRatios.reduce((a, b) => a + b, 0) / this.compressionRatios.length;
  }

  /**
   * Update deserialization metrics
   */
  private updateDeserializationMetrics(deserializationTime: number): void {
    this.metrics.totalDeserialized++;

    this.deserializationTimes.push(deserializationTime);

    if (this.deserializationTimes.length > 100) {
      this.deserializationTimes.shift();
    }

    this.metrics.averageDeserializationTime =
      this.deserializationTimes.reduce((a, b) => a + b, 0) / this.deserializationTimes.length;
  }

  /**
   * Get current serialization metrics
   */
  getMetrics(): SerializationMetrics {
    return { ...this.metrics };
  }

  /**
   * Reset metrics (useful for testing)
   */
  resetMetrics(): void {
    this.metrics = {
      totalSerialized: 0,
      totalDeserialized: 0,
      totalBytesSaved: 0,
      averageCompressionRatio: 1.0,
      averageSerializationTime: 0,
      averageDeserializationTime: 0,
      errorCount: 0
    };
    this.serializationTimes = [];
    this.deserializationTimes = [];
    this.compressionRatios = [];
  }

  /**
   * Get compression efficiency report
   */
  getEfficiencyReport(): {
    totalBytesSaved: number;
    averageCompressionRatio: number;
    averageSpeedupRatio: number;
    isPerformingWell: boolean;
    recommendations: string[];
  } {
    const recommendations: string[] = [];

    if (this.metrics.averageCompressionRatio < 1.2) {
      recommendations.push('Consider using JSON for small payloads as MessagePack overhead may not be worth it');
    }

    if (this.metrics.averageSerializationTime > 5) {
      recommendations.push('High serialization times detected - consider simplifying data structures');
    }

    if (this.metrics.errorCount > 0) {
      recommendations.push(`${this.metrics.errorCount} serialization errors detected - check data types and circular references`);
    }

    const isPerformingWell = this.metrics.averageCompressionRatio >= 1.4 &&
                           this.metrics.averageSerializationTime < 2 &&
                           this.metrics.errorCount === 0;

    return {
      totalBytesSaved: this.metrics.totalBytesSaved,
      averageCompressionRatio: this.metrics.averageCompressionRatio,
      averageSpeedupRatio: this.metrics.averageSerializationTime > 0 ?
        1 / this.metrics.averageSerializationTime : 0,
      isPerformingWell,
      recommendations
    };
  }
}